disco_aws_automation/disco_metanetwork.py

```diff
@@ -243,38 +261,88 @@ def get_interface(self, private_ip):
         Allocate a 'floating' network inteface with static ip --
         if it does not already exist.
         """
-        interface_filter = self.vpc.vpc_filter()
-        interface_filter["private-ip-address"] = private_ip
-        interfaces = self.vpc.vpc.connection.get_all_network_interfaces(
-            filters=interface_filter
+        instance_filter = self.vpc_filter()
+        instance_filter["private-ip-address"] = private_ip
+        interfaces = self._connection.get_all_network_interfaces(
+            filters=instance_filter
         )
         if interfaces:
             return interfaces[0]
 
         logging.debug("Creating floating ENI %s", private_ip)
         aws_subnet = self.subnet_by_ip(private_ip)
-        return self.vpc.vpc.connection.create_network_interface(
+        return self._connection.create_network_interface(
             subnet_id=aws_subnet['SubnetId'],
             private_ip_address=private_ip,
             description="floating interface",
-            groups=[self.security_group.id],
+            groups=[self.security_group.id]
         )
 
-    def add_sg_rule(self, protocol, ports, sg_source=None, cidr_source=None):
-        """ Add a security rule to the network """
-        sg_args = {
-            "group_id": self.security_group.id,
-            "ip_protocol": protocol
+    @staticmethod
+    def _convert_sg_rule_tuple_to_dict(sg_rule_tuple):
+        sg_rule = {
+            "group_id": sg_rule_tuple[0],
+            "ip_protocol": sg_rule_tuple[1]
         }
-        if sg_source:
-            sg_args["src_security_group_group_id"] = sg_source
-        if cidr_source:
-            sg_args["cidr_ip"] = cidr_source
+        if sg_rule_tuple[4]:
+            sg_rule["src_security_group_group_id"] = sg_rule_tuple[4]
+        elif sg_rule_tuple[5]:
+            sg_rule["cidr_ip"] = sg_rule_tuple[5]
+
+        sg_rule["from_port"] = sg_rule_tuple[2]
+        sg_rule["to_port"] = sg_rule_tuple[3]
+
+        return sg_rule
 
-        sg_args["from_port"] = ports[0]
-        sg_args["to_port"] = ports[1]
-        logging.debug("Adding sg_rule: %s", sg_args)
-        self.vpc.vpc.connection.authorize_security_group(**sg_args)
+    def create_sg_rule_tuple(self, protocol, ports, sg_source_id=None, cidr_source=None):
+        """ Creates a tuple represeting a security group rule with the security groupd ID
+        of the current meta network added """
+        return self.security_group.id, protocol, ports[0], ports[1], sg_source_id, cidr_source
+
+    def update_sg_rules(self, desire_sg_rules, dry_run=False):
```

Comment:
Should this be the only entry point for mucking with sg_rules, if so the revoke_sg_rules() and add_sg_rules() could be turned private.